File: backend/app/api/v1/webhooks.py
"""Provider webhooks — Zoom, Teams, Slack.

All endpoints verify their provider's HMAC signature and then acknowledge.
Event handling (e.g. "a Zoom recording finished, enqueue the meeting
pipeline") is dispatched into Inngest; see the migration plan Phase 5.
Until Inngest is wired up these handlers just log + ack, which is the
correct no-op behaviour (Zoom retries and we'll catch up once the queue
exists)."""

import hashlib
import hmac
import logging
import time
from collections import OrderedDict

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import JSONResponse, PlainTextResponse

from app.core.config import get_settings

logger = logging.getLogger(__name__)

router = APIRouter()

WEBHOOK_TIMESTAMP_TOLERANCE = 300  # 5 minutes


# ---------------------------------------------------------------------------
# Replay protection
#
# Without dedupe, a captured webhook can be re-played within the 5-min
# timestamp tolerance window. Cache the (timestamp, signature) pair after
# signature verification and reject repeats. Bounded LRU keeps memory
# constant. Per-process scope is fine for single-worker patterns; with
# multiple uvicorn workers, replays may slip through to a different worker
# but the receiver-side semantics (Inngest event idempotency, upserts
# keyed on provider message ids) absorb the duplicate work.
# ---------------------------------------------------------------------------
_SEEN_WEBHOOKS: OrderedDict[str, float] = OrderedDict()
_SEEN_MAX_SIZE = 10000
_SEEN_TTL_SECONDS = WEBHOOK_TIMESTAMP_TOLERANCE * 2


def _is_replay(provider: str, signature: str, timestamp: str) -> bool:
    if not signature or not timestamp:
        return False
    key = f"{provider}:{timestamp}:{signature}"
    now = time.time()
    while _SEEN_WEBHOOKS:
        oldest_key, oldest_ts = next(iter(_SEEN_WEBHOOKS.items()))
        if now - oldest_ts > _SEEN_TTL_SECONDS:
            _SEEN_WEBHOOKS.popitem(last=False)
        else:
            break
    if key in _SEEN_WEBHOOKS:
        return True
    if len(_SEEN_WEBHOOKS) >= _SEEN_MAX_SIZE:
        _SEEN_WEBHOOKS.popitem(last=False)
    _SEEN_WEBHOOKS[key] = now
    return False


def _verify_zoom_signature(request: Request, raw_body: bytes, settings) -> None:
    """Verify Zoom webhook request signature using HMAC-SHA256."""
    timestamp = request.headers.get("x-zm-request-timestamp", "")
    signature = request.headers.get("x-zm-signature", "")

    if not timestamp or not signature:
        raise HTTPException(status_code=401, detail="Missing Zoom signature headers")

    try:
        if abs(time.time() - int(timestamp)) > WEBHOOK_TIMESTAMP_TOLERANCE:
            raise HTTPException(status_code=401, detail="Webhook timestamp expired")
    except ValueError:
        raise HTTPException(status_code=401, detail="Invalid timestamp") from None

    message = f"v0:{timestamp}:{raw_body.decode()}"
    expected = "v0=" + hmac.new(
        settings.zoom_webhook_secret_token.encode(),
        message.encode(),
        hashlib.sha256,
    ).hexdigest()

    if not hmac.compare_digest(expected, signature):
        raise HTTPException(status_code=401, detail="Invalid Zoom signature")

    if _is_replay("zoom", signature, timestamp):
        logger.warning("zoom_webhook_replay_blocked timestamp=%s", timestamp)
        raise HTTPException(status_code=409, detail="Replay detected")


def _verify_slack_signature(request: Request, raw_body: bytes, settings) -> None:
    """Verify Slack webhook request signature using HMAC-SHA256."""
    timestamp = request.headers.get("X-Slack-Request-Timestamp", "")
    signature = request.headers.get("X-Slack-Signature", "")

    if not timestamp or not signature:
        raise HTTPException(status_code=401, detail="Missing Slack signature headers")

    try:
        if abs(time.time() - int(timestamp)) > WEBHOOK_TIMESTAMP_TOLERANCE:
            raise HTTPException(status_code=401, detail="Webhook timestamp expired")
    except ValueError:
        raise HTTPException(status_code=401, detail="Invalid timestamp") from None

    sig_basestring = f"v0:{timestamp}:{raw_body.decode()}"
    expected = "v0=" + hmac.new(
        settings.slack_signing_secret.encode(),
        sig_basestring.encode(),
        hashlib.sha256,
    ).hexdigest()

    if not hmac.compare_digest(expected, signature):
        raise HTTPException(status_code=401, detail="Invalid Slack signature")

    if _is_replay("slack", signature, timestamp):
        logger.warning("slack_webhook_replay_blocked timestamp=%s", timestamp)
        raise HTTPException(status_code=409, detail="Replay detected")


@router.post("/zoom")
async def zoom_webhook(
    request: Request,
) -> JSONResponse:
    """Handle Zoom webhook events (recording.completed, meeting.ended, etc.)."""
    raw_body = await request.body()
    body = await request.json()
    event = body.get("event", "")

    logger.info("Zoom webhook received: %s", event)

    settings = get_settings()

    # Verify signature for ALL requests including challenges
    _verify_zoom_signature(request, raw_body, settings)

    # Zoom URL validation challenge — must return plainToken + encryptedToken
    if event == "endpoint.url_validation":
        plain_token = body.get("payload", {}).get("plainToken", "")
        encrypted_token = hmac.new(
            settings.zoom_webhook_secret_token.encode(),
            plain_token.encode(),
            hashlib.sha256,
        ).hexdigest()
        return JSONResponse(
            content={"plainToken": plain_token, "encryptedToken": encrypted_token}
        )

    if event == "recording.completed":
        payload = body.get("payload", {}).get("object", {})
        meeting_id_str = payload.get("id", "")
        download_url = None
        for file in payload.get("recording_files", []):
            if file.get("recording_type") == "shared_screen_with_speaker_view":
                download_url = file.get("download_url")
                break
            if file.get("file_type") == "MP4":
                download_url = file.get("download_url")

        logger.info(
            "Zoom recording.completed for meeting %s, has_download=%s",
            meeting_id_str,
            bool(download_url),
        )
        if download_url:
            from app.integrations.inngest import send_event

            await send_event(
                "zoom/recording.completed",
                {
                    "zoom_meeting_id": meeting_id_str,
                    "download_url": download_url,
                    "topic": payload.get("topic"),
                },
            )

    elif event == "meeting.ended":
        logger.info("Zoom meeting.ended event received")

    return JSONResponse(content={"received": True})


@router.post("/teams", response_model=None)
async def teams_webhook(
    request: Request,
) -> JSONResponse | PlainTextResponse:
    """Handle Microsoft Teams webhook events.

    Teams uses a validation token handshake on subscription creation,
    and sends change notifications for subscribed resources.
    """
    # Teams subscription validation — must echo the validationToken as plain text
    validation_token = request.query_params.get("validationToken")
    if validation_token:
        logger.info("Teams subscription validation received")
        return PlainTextResponse(content=validation_token)

    body = await request.json()

    # Validate the client state token
    settings = get_settings()
    if not settings.teams_webhook_secret:
        raise HTTPException(status_code=500, detail="Teams webhook secret not configured")

    notifications = body.get("value", [])
    for notification in notifications:
        client_state = notification.get("clientState", "")
        if not hmac.compare_digest(client_state, settings.teams_webhook_secret):
            raise HTTPException(status_code=401, detail="Invalid Teams client state")

    for notification in notifications:
        resource = notification.get("resource", "")
        change_type = notification.get("changeType", "")

        logger.info(
            "Teams webhook: resource=%s, changeType=%s",
            resource,
            change_type,
        )

        if "communications/callRecords" in resource:
            # Resource path looks like "communications/callRecords('<id>')".
            call_record_id = ""
            if "('" in resource and "')" in resource:
                call_record_id = resource.split("('")[1].split("')")[0]
            elif "/" in resource:
                call_record_id = resource.rsplit("/", 1)[-1]

            tenant_id = notification.get("tenantId") or notification.get("tenant_id")

            if call_record_id:
                from app.integrations.inngest import send_event

                await send_event(
                    "teams/call_record.created",
                    {"call_record_id": call_record_id, "tenant_id": tenant_id},
                )
            logger.info("teams_call_record_available_for_ingest resource=%s", resource)

    return JSONResponse(content={"received": True})


@router.post("/slack/events")
async def slack_events(
    request: Request,
) -> JSONResponse:
    """Handle Slack event subscriptions.

    Slack sends a url_verification challenge during setup,
    and event callbacks for subscribed events.
    """
    raw_body = await request.body()
    body = await request.json()

    # Verify signature for ALL requests including challenges
    settings = get_settings()
    _verify_slack_signature(request, raw_body, settings)

    # Slack URL verification — return the challenge value
    if body.get("type") == "url_verification":
        logger.info("Slack URL verification challenge received")
        return JSONResponse(content={"challenge": body.get("challenge", "")})

    event = body.get("event", {})
    event_type = event.get("type", "")

    logger.info("Slack event received: %s", event_type)

    if event_type == "message":
        channel = event.get("channel", "")
        # Don't log message content for security
        logger.info("Slack message in channel %s", channel)

    elif event_type == "app_mention":
        logger.info("Slack app mention received")

    return JSONResponse(content={"received": True})


@router.post("/slack/commands")
async def slack_commands(
    request: Request,
) -> dict:
    """Handle Slack slash commands (e.g., /cognisuite status, /cognisuite meetings)."""
    raw_body = await request.body()
    settings = get_settings()
    _verify_slack_signature(request, raw_body, settings)

    form_data = await request.form()
    command = form_data.get("command", "")
    text = form_data.get("text", "")
    user_id = form_data.get("user_id", "")
    channel_id = form_data.get("channel_id", "")

    logger.info(
        "Slack command: command=%s text=%s user=%s channel=%s",
        command,
        text,
        user_id,
        channel_id,
    )

    # Parse subcommand
    parts = str(text).strip().split(maxsplit=1) if text else []
    subcommand = parts[0] if parts else "help"

    if subcommand == "help":
        return {
            "response_type": "ephemeral",
            "text": (
                "*CogniSuite Commands:*\n"
                "• `/cognisuite status` — Show recent meeting processing status\n"
                "• `/cognisuite meetings` — List recent meetings\n"
                "• `/cognisuite help` — Show this help message"
            ),
        }

    if subcommand == "status":
        return {
            "response_type": "ephemeral",
            "text": "Checking meeting processing status...",
        }

    if subcommand == "meetings":
        return {
            "response_type": "ephemeral",
            "text": "Fetching recent meetings...",
        }

    return {
        "response_type": "ephemeral",
        "text": "Unknown command. Use `/cognisuite help` for usage.",
    }

File: worker/internal/httpapi/webhooks.go
package httpapi

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/integrations/inngest"
	"github.com/go-chi/chi/v5"
)

// webhooks.go ports app/api/v1/webhooks.py: the Zoom / Teams / Slack inbound
// webhooks. Each verifies its provider's HMAC signature (or clientState) against
// the RAW body before parsing untrusted JSON, handles the provider's
// challenge/validation handshake, then fires an Inngest event for real events.
// PUBLIC — the provider signature is the capability, so no requireAuth.

const webhookTimestampTolerance = 300 // seconds (5 min)

func (s *Server) inngest() *inngest.Sender { return inngest.New(s.Cfg.InngestEventKey) }

// replayCache is a bounded, TTL'd set of (provider,timestamp,signature) tuples to
// reject replays within the timestamp tolerance window (ports _is_replay).
type replayCache struct {
	mu   sync.Mutex
	seen map[string]time.Time
	max  int
	ttl  time.Duration
}

var webhookReplay = &replayCache{seen: map[string]time.Time{}, max: 10000, ttl: 2 * webhookTimestampTolerance * time.Second}

func (c *replayCache) isReplay(provider, signature, timestamp string) bool {
	if signature == "" || timestamp == "" {
		return false
	}
	key := provider + ":" + timestamp + ":" + signature
	c.mu.Lock()
	defer c.mu.Unlock()
	now := time.Now()
	for k, t := range c.seen {
		if now.Sub(t) > c.ttl {
			delete(c.seen, k)
		}
	}
	if _, ok := c.seen[key]; ok {
		return true
	}
	if len(c.seen) >= c.max {
		for k := range c.seen { // bound memory; evict one
			delete(c.seen, k)
			break
		}
	}
	c.seen[key] = now
	return false
}

func hmacSHA256Hex(secret, msg string) string {
	m := hmac.New(sha256.New, []byte(secret))
	m.Write([]byte(msg))
	return hex.EncodeToString(m.Sum(nil))
}

func absInt64(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

// RegisterWebhooks mounts the PUBLIC provider webhooks (wired outside requireAuth).
func (s *Server) RegisterWebhooks(r chi.Router) {
	r.Post("/webhooks/zoom", s.zoomWebhook)
	r.Post("/webhooks/teams", s.teamsWebhook)
	r.Post("/webhooks/slack/events", s.slackEvents)
	r.Post("/webhooks/slack/commands", s.slackCommands)
}

// ---- signature verification (returns 0,"" when valid) ----------------------

func (s *Server) verifyZoom(r *http.Request, rawBody []byte) (int, string) {
	ts := r.Header.Get("x-zm-request-timestamp")
	sig := r.Header.Get("x-zm-signature")
	if ts == "" || sig == "" {
		return http.StatusUnauthorized, "Missing Zoom signature headers"
	}
	tsi, err := strconv.ParseInt(ts, 10, 64)
	if err != nil {
		return http.StatusUnauthorized, "Invalid timestamp"
	}
	if absInt64(time.Now().Unix()-tsi) > webhookTimestampTolerance {
		return http.StatusUnauthorized, "Webhook timestamp expired"
	}
	expected := "v0=" + hmacSHA256Hex(s.Cfg.ZoomWebhookSecretToken, "v0:"+ts+":"+string(rawBody))
	if !hmac.Equal([]byte(expected), []byte(sig)) {
		return http.StatusUnauthorized, "Invalid Zoom signature"
	}
	if webhookReplay.isReplay("zoom", sig, ts) {
		return http.StatusConflict, "Replay detected"
	}
	return 0, ""
}

func (s *Server) verifySlack(r *http.Request, rawBody []byte) (int, string) {
	ts := r.Header.Get("X-Slack-Request-Timestamp")
	sig := r.Header.Get("X-Slack-Signature")
	if ts == "" || sig == "" {
		return http.StatusUnauthorized, "Missing Slack signature headers"
	}
	tsi, err := strconv.ParseInt(ts, 10, 64)
	if err != nil {
		return http.StatusUnauthorized, "Invalid timestamp"
	}
	if absInt64(time.Now().Unix()-tsi) > webhookTimestampTolerance {
		return http.StatusUnauthorized, "Webhook timestamp expired"
	}
	expected := "v0=" + hmacSHA256Hex(s.Cfg.SlackSigningSecret, "v0:"+ts+":"+string(rawBody))
	if !hmac.Equal([]byte(expected), []byte(sig)) {
		return http.StatusUnauthorized, "Invalid Slack signature"
	}
	if webhookReplay.isReplay("slack", sig, ts) {
		return http.StatusConflict, "Replay detected"
	}
	return 0, ""
}

// ---- handlers --------------------------------------------------------------

func (s *Server) zoomWebhook(w http.ResponseWriter, r *http.Request) {
	rawBody, _ := io.ReadAll(io.LimitReader(r.Body, 4<<20))
	if code, msg := s.verifyZoom(r, rawBody); code != 0 {
		writeError(w, code, msg)
		return
	}
	var body struct {
		Event   string `json:"event"`
		Payload struct {
			PlainToken string `json:"plainToken"`
			Object     struct {
				ID             json.Number `json:"id"`
				Topic          string      `json:"topic"`
				RecordingFiles []struct {
					RecordingType string `json:"recording_type"`
					FileType      string `json:"file_type"`
					DownloadURL   string `json:"download_url"`
				} `json:"recording_files"`
			} `json:"object"`
		} `json:"payload"`
	}
	_ = json.Unmarshal(rawBody, &body)

	if body.Event == "endpoint.url_validation" {
		writeJSON(w, http.StatusOK, map[string]string{
			"plainToken":     body.Payload.PlainToken,
			"encryptedToken": hmacSHA256Hex(s.Cfg.ZoomWebhookSecretToken, body.Payload.PlainToken),
		})
		return
	}
	if body.Event == "recording.completed" {
		downloadURL := ""
		for _, f := range body.Payload.Object.RecordingFiles {
			if f.RecordingType == "shared_screen_with_speaker_view" {
				downloadURL = f.DownloadURL
				break
			}
			if f.FileType == "MP4" {
				downloadURL = f.DownloadURL
			}
		}
		if downloadURL != "" {
			s.inngest().Send(r.Context(), "zoom/recording.completed", map[string]any{
				"zoom_meeting_id": body.Payload.Object.ID.String(),
				"download_url":    downloadURL,
				"topic":           body.Payload.Object.Topic,
			})
		}
	}
	writeJSON(w, http.StatusOK, map[string]any{"received": true})
}

func (s *Server) teamsWebhook(w http.ResponseWriter, r *http.Request) {
	// Subscription validation handshake — echo validationToken as plain text.
	if vt := r.URL.Query().Get("validationToken"); vt != "" {
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(vt))
		return
	}
	rawBody, _ := io.ReadAll(io.LimitReader(r.Body, 4<<20))
	secret := s.Cfg.TeamsClientStateSecret()
	if secret == "" {
		writeError(w, http.StatusInternalServerError, "Teams webhook secret not configured")
		return
	}
	var body struct {
		Value []struct {
			ClientState string `json:"clientState"`
			Resource    string `json:"resource"`
			ChangeType  string `json:"changeType"`
			TenantID    string `json:"tenantId"`
		} `json:"value"`
	}
	_ = json.Unmarshal(rawBody, &body)

	for _, n := range body.Value {
		if !hmac.Equal([]byte(n.ClientState), []byte(secret)) {
			writeError(w, http.StatusUnauthorized, "Invalid Teams client state")
			return
		}
	}
	for _, n := range body.Value {
		if strings.Contains(n.Resource, "communications/callRecords") {
			id := extractCallRecordID(n.Resource)
			if id != "" {
				s.inngest().Send(r.Context(), "teams/call_record.created", map[string]any{
					"call_record_id": id, "tenant_id": n.TenantID,
				})
			}
		}
	}
	writeJSON(w, http.StatusOK, map[string]any{"received": true})
}

// extractCallRecordID pulls the id out of a Graph resource path like
// "communications/callRecords('<id>')" (ports the inline parse).
func extractCallRecordID(resource string) string {
	if i := strings.Index(resource, "('"); i >= 0 {
		if j := strings.Index(resource[i+2:], "')"); j >= 0 {
			return resource[i+2 : i+2+j]
		}
	}
	if i := strings.LastIndex(resource, "/"); i >= 0 {
		return resource[i+1:]
	}
	return ""
}

func (s *Server) slackEvents(w http.ResponseWriter, r *http.Request) {
	rawBody, _ := io.ReadAll(io.LimitReader(r.Body, 4<<20))
	if code, msg := s.verifySlack(r, rawBody); code != 0 {
		writeError(w, code, msg)
		return
	}
	var body struct {
		Type      string `json:"type"`
		Challenge string `json:"challenge"`
	}
	_ = json.Unmarshal(rawBody, &body)
	if body.Type == "url_verification" {
		writeJSON(w, http.StatusOK, map[string]string{"challenge": body.Challenge})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"received": true})
}

func (s *Server) slackCommands(w http.ResponseWriter, r *http.Request) {
	rawBody, _ := io.ReadAll(io.LimitReader(r.Body, 4<<20))
	if code, msg := s.verifySlack(r, rawBody); code != 0 {
		writeError(w, code, msg)
		return
	}
	form, _ := url.ParseQuery(string(rawBody))
	sub := "help"
	if fields := strings.Fields(form.Get("text")); len(fields) > 0 {
		sub = fields[0]
	}
	resp := func(text string) {
		writeJSON(w, http.StatusOK, map[string]string{"response_type": "ephemeral", "text": text})
	}
	switch sub {
	case "status":
		resp("Checking meeting processing status...")
	case "meetings":
		resp("Fetching recent meetings...")
	case "help":
		resp("*CogniSuite Commands:*\n• `/cognisuite status` — Show recent meeting processing status\n• `/cognisuite meetings` — List recent meetings\n• `/cognisuite help` — Show this help message")
	default:
		resp("Unknown command. Use `/cognisuite help` for usage.")
	}
}
